In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import cross_validate
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

In [2]:
def SelectFeature(feature_candidates,features_used,targets,df):
    N = len(feature_candidates)
    R2 = np.zeros(N)
    for k in range(N):   
        features_current = features_used.copy()
        features_current.append(feature_candidates[k])
        features = df[features_current]
        results = cross_validate(lr,features,targets,n_jobs=-1)
        R2[k] = results['test_score'].mean()
        
    R2_max = R2.max()
    feature_selected = feature_candidates[R2.argmax()] 
    return (feature_selected,R2_max)

# Use cross-validation, forward feature selection and/or feature engineering to predict the compressive strength of concrete using the features contained in the file *Concrete_train.csv* in the following ways

## (A)
### Develop a regressor that stresses interpretability

In [5]:
df = pd.read_csv('./Concrete_train.csv')
df.head()

Cement (component 1)(kg in a m^3 mixture)  \
0                                     233.81   
1                                     251.81   
2                                     252.00   
3                                     325.60   
4                                     250.00   

   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
0                                                0.0       
1                                                0.0       
2                                                0.0       
3                                              166.4       
4                                                0.0       

   Fly Ash (component 3)(kg in a m^3 mixture)  \
0                                       94.58   
1                                       99.94   
2                                        0.00   
3                                        0.00   
4                                       95.69   

   Water  (component 4)(kg in a m^3 mixture)  \
0                                     197.89   
1                                     146.14   
2                                     186.00   
3                                     174.00   
4                                     191.84   

   Superplasticizer (component 5)(kg in a m^3 mixture)  \
0                                              4.567     
1                                             12.350     
2                                              0.000     
3                                              8.900     
4                                              5.330     

   Coarse Aggregate  (component 6)(kg in a m^3 mixture)  \
0                                             947.04      
1                                            1006.00      
2                                            1111.00      
3                                             881.60      
4                                             948.90      

   Fine Aggregate (component 7)(kg in a m^3 mixture)  Age (day)  \
0                                             852.16         28   
1                                             899.76         56   
2                                             784.00          7   
3                                             790.00         28   
4                                             857.20         14   

   Concrete compressive strength(MPa, megapascals)   
0                                         22.835445  
1                                         44.140254  
2                                         11.465986  
3                                         61.235811  
4                                         24.655662

In [6]:
features = df.drop(df.columns[-1],axis=1)
targets  = df.iloc[:,-1]
features = (features - features.mean())/features.std()

In [7]:
lr = LinearRegression()
feature_candidates = list(features.columns)
features_used = []
R2 = []

while (len(feature_candidates) > 0):
    print('candidates left:',len(feature_candidates),end='')
    (feature_selected,R2_max) = SelectFeature(feature_candidates,features_used,targets,features)
    features_used.append(feature_selected)
    feature_candidates.remove(feature_selected)
    R2.append(R2_max)
    print('  selected:',feature_selected)
    
results = pd.DataFrame()
results['features'] = features_used
results['test R-squared'] = R2

print()
print(results)
ix = results['test R-squared'].idxmax()
print()
print('use features 0 to',ix,end='')
R2_max = results['test R-squared'].max()
print('    maximum test R-squared =',R2_max.round(3))

candidates left: 8

C:\Users\micha\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)


  selected: Cement (component 1)(kg in a m^3 mixture)
candidates left: 7  selected: Superplasticizer (component 5)(kg in a m^3 mixture)
candidates left: 6  selected: Age (day)
candidates left: 5  selected: Blast Furnace Slag (component 2)(kg in a m^3 mixture)
candidates left: 4  selected: Water  (component 4)(kg in a m^3 mixture)
candidates left: 3  selected: Fly Ash (component 3)(kg in a m^3 mixture)
candidates left: 2  selected: Coarse Aggregate  (component 6)(kg in a m^3 mixture)
candidates left: 1  selected: Fine Aggregate (component 7)(kg in a m^3 mixture)

                                            features  test R-squared
0          Cement (component 1)(kg in a m^3 mixture)        0.243398
1  Superplasticizer (component 5)(kg in a m^3 mix...        0.358201
2                                          Age (day)        0.477806
3  Blast Furnace Slag (component 2)(kg in a m^3 m...        0.547278
4          Water  (component 4)(kg in a m^3 mixture)        0.592691
5         Fly Ash

In [8]:
lr.fit(features[features_used],targets)
coef = pd.Series(lr.coef_,index=features.columns)
coef.sort_values(ascending=False)

Cement (component 1)(kg in a m^3 mixture)                13.159616
Water  (component 4)(kg in a m^3 mixture)                 9.806000
Fly Ash (component 3)(kg in a m^3 mixture)                7.056381
Coarse Aggregate  (component 6)(kg in a m^3 mixture)      6.446821
Age (day)                                                 2.136015
Fine Aggregate (component 7)(kg in a m^3 mixture)         2.117328
Blast Furnace Slag (component 2)(kg in a m^3 mixture)     1.644456
Superplasticizer (component 5)(kg in a m^3 mixture)      -3.031667
dtype: float64

#### We must drop the *Fine Aggregate* feature. We care about the amount of cement. We can see that the Amount of *Superplasticizer* has a negative effect on concrete strength (-3.031667).

## (B)
### Develop a regressor that stresses prediction accuracy.

In [9]:
features = df.drop(df.columns[-1],axis=1)
targets  = df.iloc[:,-1]
poly = PolynomialFeatures(2)
features_engineered = poly.fit_transform(features)
cols = poly.get_feature_names(features.columns)
features_engineered = pd.DataFrame(features_engineered,columns=cols)
features_engineered.head(3)
print('features shape =',features.shape)
print('features (engineered) shape =',features_engineered.shape)

features shape = (700, 8)
features (engineered) shape = (700, 45)


In [10]:
ix = (features_engineered.std() == 0)
drop_cols = features_engineered.columns[ix]
features_engineered = features_engineered.drop(drop_cols,axis=1)
print('features (engineered) shape =',features_engineered.shape)

features (engineered) shape = (700, 44)


In [11]:
features_engineered = (features_engineered - features_engineered.mean())/features_engineered.std()
features_engineered.describe()

Cement (component 1)(kg in a m^3 mixture)  \
count                               7.000000e+02   
mean                               -4.726378e-16   
std                                 1.000000e+00   
min                                -1.705731e+00   
25%                                -8.508288e-01   
50%                                -1.545651e-01   
75%                                 6.777092e-01   
max                                 2.520774e+00   

       Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
count                                       7.000000e+02       
mean                                        8.754902e-17       
std                                         1.000000e+00       
min                                        -8.519567e-01       
25%                                        -8.519567e-01       
50%                                        -6.182263e-01       
75%                                         8.399590e-01       
max                                         3.348178e+00       

       Fly Ash (component 3)(kg in a m^3 mixture)  \
count                                7.000000e+02   
mean                                -4.004733e-16   
std                                  1.000000e+00   
min                                 -8.645892e-01   
25%                                 -8.645892e-01   
50%                                 -8.645892e-01   
75%                                  9.587934e-01   
max                                  2.217792e+00   

       Water  (component 4)(kg in a m^3 mixture)  \
count                               7.000000e+02   
mean                                4.365793e-15   
std                                 1.000000e+00   
min                                -2.753839e+00   
25%                                -7.748583e-01   
50%                                 1.469845e-01   
75%                                 5.118232e-01   
max                                 2.990480e+00   

       Superplasticizer (component 5)(kg in a m^3 mixture)  \
count                                       7.000000e+02     
mean                                        1.297375e-15     
std                                         1.000000e+00     
min                                        -1.065575e+00     
25%                                        -1.065575e+00     
50%                                         8.660600e-02     
75%                                         6.643760e-01     
max                                         4.342621e+00     

       Coarse Aggregate  (component 6)(kg in a m^3 mixture)  \
count                                       7.000000e+02      
mean                                        2.245410e-14      
std                                         1.000000e+00      
min                                        -2.179938e+00      
25%                                        -5.043832e-01      
50%                                        -5.569256e-02      
75%                                         7.318183e-01      
max                                         2.083134e+00      

       Fine Aggregate (component 7)(kg in a m^3 mixture)     Age (day)  \
count                                       7.000000e+02  7.000000e+02   
mean                                        2.258194e-15  1.007131e-17   
std                                         1.000000e+00  1.000000e+00   
min                                        -2.196943e+00 -6.991254e-01   
25%                                        -5.498518e-01 -5.223177e-01   
50%                                         6.954236e-02 -2.972896e-01   
75%                                         6.343291e-01  1.527665e-01   
max                                         2.677981e+00  5.119457e+00   

       Cement (component 1)(kg in a m^3 mixture)^2  \
count                                 7.000000e+02   
mean                                  5.900042e-17   
std                                   1.000000e+00   
min 

In [12]:
lr = LinearRegression()
feature_candidates = list(features_engineered.columns)
features_used = []
R2 = []

while (len(feature_candidates) > 0):
    print('candidates left:',len(feature_candidates),end='')
    (feature_selected,R2_max) = SelectFeature(feature_candidates,features_used,targets,features_engineered)
    features_used.append(feature_selected)
    feature_candidates.remove(feature_selected)
    R2.append(R2_max)
    print('  selected:',feature_selected)
    
results = pd.DataFrame()
results['features'] = features_used
results['test R-squared'] = R2

print()
print(results)
ix = results['test R-squared'].idxmax()
print()
print('use features 0 to',ix,end='')
R2_max = results['test R-squared'].max()
print('    maximum test R-squared =',R2_max.round(3))

candidates left: 44

C:\Users\micha\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)


  selected: Superplasticizer (component 5)(kg in a m^3 mixture) Age (day)
candidates left: 43  selected: Cement (component 1)(kg in a m^3 mixture)
candidates left: 42  selected: Blast Furnace Slag (component 2)(kg in a m^3 mixture) Coarse Aggregate  (component 6)(kg in a m^3 mixture)
candidates left: 41  selected: Fly Ash (component 3)(kg in a m^3 mixture) Age (day)
candidates left: 40  selected: Coarse Aggregate  (component 6)(kg in a m^3 mixture) Age (day)
candidates left: 39  selected: Water  (component 4)(kg in a m^3 mixture)
candidates left: 38  selected: Age (day)^2
candidates left: 37  selected: Cement (component 1)(kg in a m^3 mixture) Fly Ash (component 3)(kg in a m^3 mixture)
candidates left: 36  selected: Blast Furnace Slag (component 2)(kg in a m^3 mixture) Age (day)
candidates left: 35  selected: Blast Furnace Slag (component 2)(kg in a m^3 mixture) Superplasticizer (component 5)(kg in a m^3 mixture)
candidates left: 34  selected: Superplasticizer (component 5)(kg in a m^3

#### The engineered features increase the R-squared: 0.627 (from previous loop) to 0.796

In [13]:
lr.fit(features_engineered[features_used],targets)
coef = pd.Series(lr.coef_,index=features_engineered.columns)
coef.sort_values(ascending=False)

Fly Ash (component 3)(kg in a m^3 mixture) Superplasticizer (component 5)(kg in a m^3 mixture)                273.938489
Coarse Aggregate  (component 6)(kg in a m^3 mixture)                                                          260.490521
Blast Furnace Slag (component 2)(kg in a m^3 mixture)                                                         252.501233
Coarse Aggregate  (component 6)(kg in a m^3 mixture)^2                                                        223.032798
Water  (component 4)(kg in a m^3 mixture) Fine Aggregate (component 7)(kg in a m^3 mixture)                   136.899941
Fine Aggregate (component 7)(kg in a m^3 mixture) Age (day)                                                   116.770381
Blast Furnace Slag (component 2)(kg in a m^3 mixture) Coarse Aggregate  (component 6)(kg in a m^3 mixture)     35.257012
Blast Furnace Slag (component 2)(kg in a m^3 mixture)^2                                                        15.592522
Blast Furnace Slag (component 2)